File: elixir/binary-search-tree/binary_search_tree.exs
defmodule BinarySearchTree do
  @type bst_node :: %{data: any, left: bst_node | nil, right: bst_node | nil}

  @doc """
  Create a new Binary Search Tree with root's value as the given 'data'
  """
  @spec new(any) :: bst_node
  def new(data) do
    %{data: data, left: nil, right: nil}
  end

  @doc """
  Creates and inserts a node with its value as 'data' into the tree.
  """
  @spec insert(bst_node, any) :: bst_node
  def insert(nil, data) do
    %{data: data, left: nil, right: nil}
  end
  def insert(%{data: node_data} = tree_node, data) do
    cond do
      data <= node_data ->
        %{tree_node | left: insert(tree_node.left, data)}
      data > node_data ->
        %{tree_node | right: insert(tree_node.right, data)}
    end
  end

  @doc """
  Traverses the Binary Search Tree in order and returns a list of each node's data.
  """
  @spec in_order(bst_node) :: [any]
  def in_order(tree), do: in_order([], tree)
  def in_order(acc, nil), do: acc
  def in_order(acc, %{data: data, left: l, right: r}) do
    in_order(acc, l)
    |> Kernel.++([data])
    |> in_order(r)
  end
end

File: python/binary-search-tree/binary_search_tree.py
class TreeNode(object):
    def __init__(self, data=None, left=None, right=None):
        self.data = data
        self.left = left
        self.right = right

    def __repr__(self):
        fmt = 'TreeNode(data={}, left={}, right={})'
        return fmt.format(self.data, self.left, self.right)

    def insert(self, data):
        if data <= self.data:
            if self.left:
                self.left.insert(data)
            else:
                self.left = TreeNode(data)
        elif data > self.data:
            if self.right:
                self.right.insert(data)
            else:
                self.right = TreeNode(data)

    def get_sorted(self, sorted_data=[]):
        if self.left:
            self.left.get_sorted(sorted_data)
        sorted_data.append(self.data)
        if self.right:
            self.right.get_sorted(sorted_data)
        return sorted_data


class BinarySearchTree(object):
    def __init__(self, tree_data):
        self.root = TreeNode(tree_data[0])
        for data in tree_data[1:]:
            self.root.insert(data)

    def data(self):
        return self.root

    def sorted_data(self):
        return self.root.get_sorted([])
